crates/krun-server/src/bin/krun-server.rs

```diff
@@ -1,32 +1,83 @@
 use std::net::TcpListener;
-use std::os::fd::AsRawFd;
-use std::panic;
+use std::os::unix::process::ExitStatusExt as _;
 use std::process::Command;
+use std::sync::mpsc::{channel, TryRecvError};
+use std::sync::{Arc, Condvar, Mutex};
 
 use anyhow::{Context, Result};
 use krun_server::cli_options::options;
 use krun_server::server::start_server;
-use nix::sys::socket::{shutdown, Shutdown};
+use log::debug;
 
 fn main() -> Result<()> {
     env_logger::init();
 
     let options = options().run();
 
     let listener = TcpListener::bind(format!("0.0.0.0:{}", options.server_port))?;
-    let listener_fd = listener.as_raw_fd();
+    let (child_tx, child_rx) = channel();
+    let idle_pair = Arc::new((Mutex::new(true), Condvar::new()));
 
-    let server_thread = start_server(listener);
+    start_server(listener, child_tx, Arc::clone(&idle_pair));
 
-    Command::new(&options.command)
+    let status = Command::new(&options.command)
         .args(options.command_args)
         .status()
-        .with_context(|| format!("Failed to execute command {:?}", options.command))?;
-
-    shutdown(listener_fd, Shutdown::Both)?;
-    if let Err(err) = server_thread.join() {
-        panic::resume_unwind(err);
+        .with_context(|| format!("Failed to execute {:?} as child process", options.command))?;
+    if !status.success() {
+        if let Some(code) = status.code() {
+            eprintln!(
+                "{:?} process exited with status code: {code}",
+                options.command
+            );
+        } else {
+            eprintln!(
+                "{:?} process terminated by signal: {}",
+                options.command,
+                status
+                    .signal()
+                    .expect("either one of status code or signal should be set")
+            );
+        }
     }
 
-    Ok(())
+    println!("Waiting for other commands launched through this krun server to exit...");
+    println!("Press Ctrl+C to force quit");
+
+    let (idle_lock, idle_cvar) = &*idle_pair;
+
+    loop {
+        let mut child = {
+            let _guard = idle_cvar
+                .wait_while(idle_lock.lock().unwrap(), |idle| !*idle)
+                .unwrap();
+            match child_rx.try_recv() {
+                Ok(child) => child,
+                Err(TryRecvError::Empty) => {
+                    // Server is idle and no more child processes. We're done.
+                    return Ok(());
+                },
+                Err(TryRecvError::Disconnected) => {
+                    panic!("child_tx in server thread should not be gone");
+                },
+            }
+        };
+        let child_pid = child.id();
+        let status = child
+            .wait()
+            .context("Failed to wait for child process to exit")?;
```

Comment:
Hmm... Maybe we shouldn't propagate this error. It's more important to wait for the other child processes, rather than dying here...

Reply:
Done. Made the failure to run main command not error as well, since we might already have child processes launched / being launched at that point.